src/main/kotlin/org/rust/ide/intentions/ExtractEnumVariantIntention.kt

```diff
@@ -0,0 +1,210 @@
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiDocumentManager
+import com.intellij.psi.PsiElement
+import com.intellij.psi.search.searches.ReferencesSearch
+import org.rust.ide.refactoring.RsInPlaceVariableIntroducer
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.RsVisibility
+import org.rust.lang.core.psi.ext.ancestorStrict
+import org.rust.lang.core.psi.ext.parentEnum
+import org.rust.lang.core.psi.ext.visibility
+import org.rust.lang.core.types.infer.TypeVisitor
+import org.rust.lang.core.types.infer.needToSubstitute
+import org.rust.lang.core.types.regions.ReEarlyBound
+import org.rust.lang.core.types.regions.Region
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.ty.TyReference
+import org.rust.lang.core.types.ty.TyTypeParameter
+import org.rust.lang.core.types.type
+
+class ExtractEnumVariantIntention : RsElementBaseIntentionAction<ExtractEnumVariantIntention.Context>() {
+    override fun getText() = "Extract enum variant"
+    override fun getFamilyName() = text
+
+    class Context(val variant: RsEnumVariant)
+
+    override fun findApplicableContext(project: Project, editor: Editor, element: PsiElement): Context? {
+        val enum = element.ancestorStrict<RsEnumVariant>() ?: return null
+        if (enum.tupleFields == null && enum.blockFields == null) {
+            return null
+        }
+        if (enum.variantDiscriminant != null) {
+            return null
+        }
+        return Context(enum)
+    }
+
+    override fun invoke(project: Project, editor: Editor, ctx: Context) {
+        val factory = RsPsiFactory(project)
+        val element = createElement(ctx.variant, factory)
+
+        val enum = ctx.variant.parentEnum
+        val typeParameters = filterTypeParameters(element.typeReferences, enum.typeParameterList)
+        val whereClause = enum.whereClause
+        val name = ctx.variant.name ?: return
+
+        val struct = element.createStruct(name, typeParameters, whereClause?.text ?: "")
+        val inserted = enum.parent.addBefore(struct, enum) as RsStructItem
+        val visibility = element.visibility
+
+        for (usage in ReferencesSearch
+            .search(ctx.variant)) {
+            element.replaceUsage(usage.element, name)
+        }
+
+        val newVariant = factory.createTupleFields(listOf(RsPsiFactory.TupleField(visibility, inserted.declaredType)))
+        val replaced = element.toBeReplaced.replace(newVariant)
+
+        offerStructRename(project, editor, inserted, replaced)
+    }
+
+    private fun filterTypeParameters(references: List<RsTypeReference>, parameters: RsTypeParameterList?): String {
+        return if (parameters != null) {
+            val lifetimes = parameters.lifetimeParameterList.filter { lifetime -> references.any { matchesLifetime(it, lifetime) } }
+            val typeParameters = parameters.typeParameterList.filter { param -> references.any { matchesParameter(it, param) } }
```

Comment:
I need to do O(n^2) traversals I think, since I can't simply put any side into a hashmap, I traverse them dynamically. Or not? I could gather all the types into a set and then traverse the set, but that seems too complicated.

Reply:
Regarding const generics, their type renderer generate code like this:
```
trait Trait;
struct V1<const T: usize> { a: [u32; T] }

enum A<const T: usize> {
    V1(V1<{ T }>),
    V2
}
```
I would need to create a new type renderer which disables this. The type renderer system seems a bit annoying, since I would need to create a new type renderer for each combinations of parameters that I need. What should I do here?

Reply:
> I could gather all the types into a set and then traverse the set, but that seems too complicated.
Ok, let's use the simple solution since.

Can you explain why `V1(V1<{ T }>)` is wrong? `V1(V1<T>)` doesn't compile.

Reply:
Oops, I didn't know this syntax. Although `V1(V1<T>)` seems to compile too:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=dd8e2fccc2f6ef81b29e8ef668b4fc17

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0146a7c7e27de94671e7ab15e113e0c5

Nevermind then.